query/src/main/java/org/infinispan/query/impl/IndexStartupRunner.java

```diff
@@ -55,6 +58,46 @@ private static IndexStartupMode computeFinalMode(Configuration configuration) {
             }
          }
       }
+
+      if (DataKind.SHARED_MEMORY_MAX_COUNT.equals(dataKind) && !indexesAreVolatile &&
+            IndexStartupMode.AUTO.equals(startupMode)) {
+         // in this case a node that is crashed could have not aligned indexes.
+         // we force the reindex when it recovers to be sure that is aligned.
+         return IndexStartupMode.REINDEX;
+      }
+
       return (IndexStartupMode.AUTO.equals(startupMode)) ? IndexStartupMode.NONE : startupMode;
    }
+
+   private enum DataKind {
+      VOLATILE, PERSISTENT, SHARED_MEMORY_MAX_COUNT
+   }
+
+   private static DataKind computeDataKind(Configuration configuration) {
+      List<StoreConfiguration> cacheStores = configuration.persistence().stores();
+      if (cacheStores.isEmpty()) {
+         return DataKind.VOLATILE;
+      }
+
+      boolean nonSharedCacheStore = false;
```

Comment:
isn't that superfluos?
sharedCacheStore=TRUE means that nonSharedCacheStore must be FALSE !!

Reply:
Not sure we can remove it. Since if we have 2 cache stores (one shared one not shared - I don't know if it is possible). In this case should return `DataKind.PERSISTENT`

Reply:
I would think that it can be simple 
If one is shared => must be true
Or
If one is unshared => must be false
So at least only 2 possibilities or am I missing a third case where you need to decide different?